src/cpp/session/modules/SessionGit.cpp

```diff
@@ -204,10 +204,52 @@ std::string gitBin()
 }
 #endif
 
+bool waitForIndexLock(const FilePath& workingDir)
+{
+   using namespace boost::posix_time;
+   
+   // count the number of retries (avoid getting stuck in wait loops when
+   // an index.lock file exists and is never cleaned up)
+   static int retryCount = 0;
+   
+   FilePath lockPath = workingDir.childPath(".git/index.lock");
+   
+   // wait 1 second for an 'index.lock' file to be removed
+   for (std::size_t i = 0; i < 10; ++i)
+   {
+      // if there's no lockfile, we can proceed
+      if (!lockPath.exists())
```

Comment:
One thing to be aware of here is that NFS can cache directory listings for distressingly long periods of time. If this turns out to be a cause of these `index.lock` issues, another angle of attack is to attempt to clear the NFS cache for the file. Just for your back pocket for now, but here's the routine if you need it: https://github.com/rstudio/rstudio/blob/54cd3abcfc58837b433464c793fe9b03a87f0bb4/src/cpp/core/system/PosixNfs.cpp#L29-L36